sunscreen/src/compiler.rs

```diff
@@ -213,51 +205,30 @@ impl<B> CompilerData<B> {
         match self {
             Self::Fhe(d) => d,
             Self::FheZkp(d, _) => d,
-            _ => unreachable!(),
         }
     }
 
     fn fhe_data(&self) -> &FheCompilerData {
         match self {
             Self::Fhe(d) => d,
             Self::FheZkp(d, _) => d,
-            _ => unreachable!(),
         }
     }
 
-    fn unwrap_zkp(self) -> ZkpCompilerData<B> {
-        match self {
-            Self::Zkp(d) => d,
-            Self::FheZkp(_, d) => d,
-            _ => unreachable!(),
-        }
-    }
-
-    fn unwrap_fhe(self) -> FheCompilerData {
+    fn take_fhe_data(self) -> FheCompilerData {
         match self {
             Self::Fhe(d) => d,
             Self::FheZkp(d, _) => d,
-            _ => unreachable!(),
         }
     }
 }
 
 /// A compiler for Sunscreen FHE programs.
 //
-// A note for developer sanity below: the type strategy here is to have `.fhe_program` enhance to
-// an FHE-capable compiler, and `.zkp_backend` to enhance to a ZKP-capable compiler. For that
-// strategy alone, the only methods that need to be defined in multiple impls are the
-// aforementioned ones, as the returned compiler type will change depending on the current compiler
-// type.
+// Here the markers have the following meanings:
 //
-// But there's one added complication in that the ZKP method `.zkp_program` has type restrictions
-// depending on if the current ZKP-capable compiler is also FHE-capable, so that method also needs
-// multiple definitions.
-//
-// TODO: allow FHE params to be specified for both FHE and ZKP programs. This is
-// surprisingly complicated with the current approach. We may have to split into two markers (one
-// for FHE and one for ZKP) for that to be feasible, without requiring 6 definitions for the same
-// method.
+// - FHE: indicates params are derivable, either through FHE programs or manual params setting.
+// - ZKP: indicates a ZKP backend has been specified, thus user can add ZKP programs.
```

Comment:
Should this comment be FHEZKP instead of ZKP? Since `CompilerData` specifies B which can be `Fhe` or `FheZkp`

Reply:
Actually no, sorry this is maybe a tad more confusing now. This comment is referring to the meanings of the marker traits, not the shape of the compiler data.

Reply:
Maybe it would be a little cleaner if we kept the `CompilerData::{None, Zkp}` variants, but pulled out the `FheCompilerData { params_mode }` field so that it is always present. Then the shape of the compiler data would match the marker more closely.